ocis-pkg/sync/mutex.go

```diff
@@ -0,0 +1,74 @@
+package sync
+
+import (
+	"sync"
+)
+
+// NRWMutex works the same as RWMutex, the only difference is that it stores mutexes in a map and reuses them.
+// It's handy if you want to write-lock, write-unlock, read-lock and read-unlock for specific names only.
+type NRWMutex struct {
+	m  sync.Mutex
+	mm map[string]*nrw
+}
+
+type nrw struct {
+	m sync.RWMutex
+	c int
+}
+
+// NewNRWMutex returns a new instance of NRWMutex.
+func NewNRWMutex() NRWMutex {
+	return NRWMutex{mm: make(map[string]*nrw)}
+}
+
+// Lock locks rw for writing.
+func (c *NRWMutex) Lock(k string) {
+	c.m.Lock()
```

Comment:
this is the top level locking I don't understand. Imagine the case of 3 goroutines (a, b and c) fighting for the same resource, in this case an instance of a `NRWMutex`; `b` executes faster and manages to gain access to the shared resource, and **lets a & c excluded to modify anything from the critical section** (the CS being the map of rwmutexes). Then `b` goes ahead and does its thing, all this while a & c are waiting for their turn. But essentially what `b` does is getting access to yet another resource that is critically excluded from all the others.

What I'm saying is that, if a map of `RWMutex` is already preventing goroutines from accessing shared resources in a chaotic manner, why does access to this list (the `map[string]sync.RWMutex`) has to be a critical section?

Imagine that `a` & `c` want to access to accounts `1` and `3` respectively, and `b` wants access to account `2`, why do a and c have to be waiting while b does something on a resource that goroutines a and c don't really need to read or write into? This would make sense if we would have to keep foreign indexes between accounts, in that case I do understand a higher level locking, sort of like a "transaction".